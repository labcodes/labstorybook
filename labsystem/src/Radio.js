/* eslint-disable react/require-default-props */
import React from "react";
import PropTypes, { string, number, bool } from "prop-types";
import { isUndefined } from "lodash";

// Radio //

export default class Radio extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([string, number, bool]).isRequired,
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    className: PropTypes.string,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    disabled: false,
    checked: undefined,
    defaultChecked: undefined,
    className: undefined,
    onChange: () => {},
  };

  constructor(props) {
    super(props);
    const { defaultChecked, checked, id } = props;
    if (!isUndefined(defaultChecked) && !isUndefined(checked)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both checked and defaultChecked for radio input ${id} at the same time. We always initialize the radio with defaultChecked. Make sure this is the behaviour you want.`
      );
    }

    let localChecked = false;
    if (defaultChecked) {
      localChecked = defaultChecked;
    } else if (!isUndefined(checked)) {
      localChecked = checked;
    }

    this.state = {
      localChecked,
    };
  }

  componentDidUpdate(prevProps) {
    const { checked } = this.props;
    if (checked !== prevProps.checked) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState(() => ({ localChecked: checked }));
    }
  }

  handleOnChange = (event) => {
    const { onChange } = this.props;
    if (!isUndefined(onChange)) {
      onChange(event);
    }

    this.setState((state) => ({ localChecked: !state.localChecked }));
  };

  render() {
    const { className, id, label, disabled, checked, name, value } = this.props;

    return (
      <React.Fragment>
        <input
          className={`lab-radio ${className || ""}`}
          type="radio"
          id={id}
          {...(disabled ? { disabled } : undefined)}
          checked={checked}
          name={name}
          value={value}
          onChange={this.handleOnChange}
        />
        <label className="lab-radio__label" htmlFor={id}>
          <span className="lab-radio__container" />
          {label}
        </label>
      </React.Fragment>
    );
  }
}
