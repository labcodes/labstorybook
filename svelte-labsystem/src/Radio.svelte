<script>
  export let id;
  export let name;
  export let label;
  export let group = undefined;
  export let value = undefined;
  export let disabled = undefined;

  let className = "";
  export { className as class };
</script>

<input
  class={`lab-radio${` ${className}` || ''}`}
  type="radio"
  {id}
  {name}
  {value}
  {disabled}
  bind:group
  on:change />

<label class="lab-radio__label" for={id}>
  <span class="lab-radio__container" />
  {label}
</label>
