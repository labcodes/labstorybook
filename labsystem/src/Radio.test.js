/* eslint-disable no-console */
import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import Radio from "./Radio";

describe("Radio", () => {
  const originalWarn = console.warn;
  afterEach(() => {
    console.warn = originalWarn;
  });

  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(
        <Radio
          name="test-radio-group"
          id="test-radio"
          label="test radio"
          value="radio1"
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("raises console.warn when passing checked and defaultChecked at the same time", async () => {
    console.warn = jest.fn();

    shallow(
      <Radio
        name="test-radio-group"
        id="test-radio"
        label="test radio"
        value="radio1"
        checked
        defaultChecked
      />
    );

    const consoleText =
      "You are setting both checked and defaultChecked for radio input test-radio at the same time. We always initialize the radio with defaultChecked. Make sure this is the behaviour you want.";

    expect(console.warn).toBeCalledWith(consoleText);
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(
        <Radio
          name="test-radio"
          id="test-radio"
          label="test radio"
          value="radio1"
          disabled
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a different className", async () => {
    const renderedComponent = renderer
      .create(
        <Radio
          name="test-radio"
          id="test-radio"
          label="test radio"
          value="radio1"
          className="lab-radio--purple"
          checked
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("inits state.localChecked with defaultChecked if defined", async () => {
    let component = shallow(
      <Radio
        name="test-radio"
        id="test-radio"
        label="test radio"
        value="radio1"
        defaultChecked
      />
    );
    expect(component.state().localChecked).toBe(true);

    component = shallow(
      <Radio
        name="test-radio"
        id="test-radio"
        label="test radio"
        value="radio1"
        defaultChecked={false}
      />
    );
    expect(component.state().localChecked).toBe(false);

    component = shallow(
      <Radio
        name="test-radio"
        id="test-radio"
        label="test radio"
        value="radio1"
      />
    );
    expect(component.state().localChecked).toBe(false);
  });

  it("changes state when input changes", async () => {
    const component = shallow(
      <Radio
        name="test-radio"
        id="test-radio"
        label="test radio"
        value="radio1"
      />
    );

    expect(component.state().localChecked).toBe(false);
    component.find("input").at(0).simulate("change");
    expect(component.state().localChecked).toBe(true);
  });

  it("calls props.onChange passing event when input changes", async () => {
    const mockOnChange = jest.fn();
    const component = shallow(
      <Radio
        name="test-radio"
        id="test-radio"
        label="test radio"
        value="radio1"
        onChange={mockOnChange}
      />
    );

    expect(component.state().localChecked).toBe(false);
    expect(mockOnChange).not.toBeCalled();

    component.find("input").at(0).simulate("change", { test: "event" });

    expect(component.state().localChecked).toBe(true);
    expect(mockOnChange).toBeCalledWith({ test: "event" });
  });
});
