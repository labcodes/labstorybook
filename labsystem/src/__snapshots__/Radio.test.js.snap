// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders as expected when passing a different className 1`] = `
<div>
  <input
    checked={true}
    className="lab-radio lab-radio--purple"
    disabled={false}
    id="test-radio"
    name="test-radio"
    onChange={[Function]}
    type="radio"
    value="radio1"
  />
  <label
    className="lab-radio__label"
    htmlFor="test-radio"
  >
    <span
      className="lab-radio__container"
    />
    test radio
  </label>
</div>
`;

exports[`Radio renders as expected when passing disabled as true 1`] = `
<div>
  <input
    className="lab-radio "
    disabled={true}
    id="test-radio"
    name="test-radio"
    onChange={[Function]}
    type="radio"
    value="radio1"
  />
  <label
    className="lab-radio__label"
    htmlFor="test-radio"
  >
    <span
      className="lab-radio__container"
    />
    test radio
  </label>
</div>
`;

exports[`Radio renders with base props 1`] = `
<div>
  <input
    className="lab-radio "
    disabled={false}
    id="test-radio"
    name="test-radio-group"
    onChange={[Function]}
    type="radio"
    value="radio1"
  />
  <label
    className="lab-radio__label"
    htmlFor="test-radio"
  >
    <span
      className="lab-radio__container"
    />
    test radio
  </label>
</div>
`;
